### PR TITLE
Add support for advanced SSM parameters

### PR DESCRIPTION
**Pull request by davidcelis**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
resource/aws_ssm_parameter: Add `tier` argument [GH-8525]
```

### CODE REVIEW

File: aws/resource_aws_ssm_parameter_test.go [hunk 1]
```diff
@@ -53,6 +53,33 @@ func TestAccAWSSSMParameter_basic(t *testing.T) {
 						regexp.MustCompile(fmt.Sprintf("^arn:aws:ssm:[a-z0-9-]+:[0-9]{12}:parameter/%s$", name))),
 					resource.TestCheckResourceAttr("aws_ssm_parameter.foo", "value", "bar"),
 					resource.TestCheckResourceAttr("aws_ssm_parameter.foo", "type", "String"),
+					resource.TestCheckResourceAttr("aws_ssm_parameter.foo", "tier", "Basic"),
```
[the remark this answers is not in the document]
Comment (davidcelis) [reply aws/resource_aws_ssm_parameter_test.go]: 🤦‍♂ yup!! I thought I had caught all of my misnaming, but I must have missed this one 😂

**Comment by bflad** [comment]
Hey @davidcelis, thanks for the updates here! When testing this for merging I swapped out the testing to also perform updates for the `tier` argument similar to https://github.com/terraform-providers/terraform-provider-aws/pull/8525#discussion_r281366851 and discovered that updating from `Advanced` to `Standard` returned the following API error:

```
--- FAIL: TestAccAWSSSMParameter_Tier (25.65s)
    testing.go:568: Step 1 error: errors during apply:

        Error: error creating SSM parameter: ValidationException: This parameter uses the advanced-parameter tier. You can't downgrade a parameter from the advanced-parameter tier to the standard-parameter tier. If necessary, you can delete the advanced parameter and recreate it as a standard parameter. Be aware that standard parameters have a value limit of 4096 characters.
        	status code: 400, request id: 8ffb05d8-5977-4226-8ef5-276b5be1fece
```

So! Turns out this needed some extra love for that specific scenario and in this case a relatively safe and simple `CustomizeDiff` only trigger resource recreation for this particular issue:

```go
		CustomizeDiff: customdiff.All(
			// Prevent the following error during tier update from Advanced to Standard:
			// ValidationException: This parameter uses the advanced-parameter tier. You can't downgrade a parameter from the advanced-parameter tier to the standard-parameter tier. If necessary, you can delete the advanced parameter and recreate it as a standard parameter.
			customdiff.ForceNewIfChange("tier", func(old, new, meta interface{}) bool {
				return old.(string) == ssm.ParameterTierAdvanced && new.(string) == ssm.ParameterTierStandard
			}),
		),
```

With this in place, everything looks good. 🚀  Thanks so much for your contribution here! Should be able to get this released today. 👍 

Output from acceptance testing after above changes:

```
--- PASS: TestAccAWSSSMParameter_disappears (9.99s)
--- PASS: TestAccAWSSSMParameter_basic (13.47s)
--- PASS: TestAccAWSSSMParameter_secure (13.74s)
--- PASS: TestAccAWSSSMParameter_fullPath (19.63s)
--- PASS: TestAccAWSSSMParameter_importBasic (21.78s)
--- PASS: TestAccAWSSSMParameter_updateTags (26.12s)
--- PASS: TestAccAWSSSMParameter_changeNameForcesNew (30.26s)
--- PASS: TestAccAWSSSMParameter_overwrite (32.02s)
--- PASS: TestAccAWSSSMParameter_updateDescription (35.51s)
--- PASS: TestAccAWSSSMParameter_Tier (37.83s)
--- PASS: TestAccAWSSSMParameter_secure_with_key (46.98s)
--- PASS: TestAccAWSSSMParameter_secure_keyUpdate (61.83s)
```

**Comment by davidcelis** [comment]
oh wow, thank you for helping me bring this across the finish line, @bflad!!

**Comment by bflad** [comment]
This has been released in [version 2.10.0 of the Terraform AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#2100-may-10-2019). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

**Comment by marpada** [comment]
@davidcelis thank you for being awesome!

**Comment by smolse** [comment]
Hi guys. I'm afraid this pull request breaks SSM parameters management in AWS China. Please check #8662.

**Pull request merged**